pkg/platform/runtime/buildplan/changeset.go

```diff
@@ -32,3 +32,46 @@ func NewBaseArtifactChangesetByBuildPlan(build *model.Build, requestedOnly, buil
 
 	return artifact.NewArtifactChangeset(nil, new, requestedOnly), nil
 }
+
+func TopLevelArtifactsAdded(changeset artifact.ArtifactChangeset, artifacts artifact.Map) []artifact.ArtifactID {
+	var added []artifact.ArtifactID
+	for _, candidate := range changeset.Added {
+		if !IsDependency(candidate.ArtifactID, changeset, artifacts) {
+			foundId := candidate.ArtifactID
+			added = append(added, foundId)
+		}
+	}
+	return added
+}
+
+// IsDependency iterates over all artifacts and their dependencies in the given changeset, and
+// returns whether or not the given artifact is a dependency of any of those artifacts or
+// dependencies.
+func IsDependency(a artifact.ArtifactID, changeset artifact.ArtifactChangeset, artifacts artifact.Map) bool {
+	for _, artifact := range changeset.Added {
+		if artifact.ArtifactID == a {
+			continue
+		}
+
+		for _, depId := range RecursiveDependenciesFor(artifact.ArtifactID, artifacts) {
```

Comment:
I know this isn't related to your PR but I took a look at the `RecursiveDependenciesFor` function and it is only grabbing direct dependencies, not dependencies of the steps that generated the given artifact. Is this something we care to include?

Reply:
Not sure I follow your meaning, it seems recursive? https://github.com/ActiveState/cli/blob/DX-2556/pkg/platform/runtime/buildplan/dependencies.go#L181-L181

Reply:
`artf.Dependencies` is only the direct dependencies of the artifact. If we wanted the full closure we would recursively find the step that generated the artifact and include the dependences of that step.

Reply:
That might be overkill for what we want though.